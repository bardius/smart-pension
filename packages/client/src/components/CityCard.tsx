import React, { useCallback } from 'react'
import type { VFC } from 'react'
import { Box, Text, IconButton } from '@chakra-ui/react'
import { PlusSquareIcon, StarIcon } from '@chakra-ui/icons'
import { useUpdateCity } from '../contexts/cities/hooks'

export type CityCardProps = {
  id: number
  title: string
  subTitle: string
  visited: boolean
  wishlist: boolean
}

export const CityCard: VFC<CityCardProps> = props => {
  const { id, title, subTitle, visited, wishlist } = props
  // This could be moved to a prop and drill it down in order to make this card more generic
  const updateCity = useUpdateCity()

  const onToggleVisited = useCallback(() => {
    updateCity({
      variables: { input: { id: id, visited: !visited, wishlist: wishlist } },
      optimisticResponse: {
        updateCity: {
          id: id,
          name: title,
          country: subTitle,
          visited: !visited,
          wishlist: wishlist,
        },
      },
    })
  }, [updateCity, id, visited, wishlist, subTitle, title])

  const onToggleWishlist = useCallback(() => {
    updateCity({
      variables: { input: { id: id, visited: visited, wishlist: !wishlist } },
      optimisticResponse: {
        updateCity: {
          id: id,
          name: title,
          country: subTitle,
          visited: visited,
          wishlist: !wishlist,
        },
      },
    })
  }, [updateCity, id, visited, wishlist, subTitle, title])

  const wishlistBtnLabel = `${wishlist ? 'Added in wishlist' : 'Not in wishlist'}, Click to toggle status`
  const visitedBtnLabel = `${visited ? 'Marked as visited' : 'Marked as not visited'}, Click to toggle status`

  return (
    <Box
      as="li"
      p={4}
      display="flex"
      flexDirection={{ sm: 'column', lg: 'row' }}
      borderWidth={1}
      data-testid="city-card"
    >
      <Box textAlign={'left'} maxWidth={220}>
        <Text fontWeight="bold" fontSize="lg">
          {title}
        </Text>
        <Text as="span" fontWeight="normal" fontSize="md">
          ({subTitle})
        </Text>
      </Box>
      <Box ml={'auto'}>
        <IconButton
          data-testid={`city-card-visited-btn${visited ? '-active' : ''}`}
          mx={1}
          variant="outline"
          colorScheme={visited ? 'blue' : 'red'}
          title={visitedBtnLabel}
          aria-label={visitedBtnLabel}
          onClick={onToggleVisited}
          icon={<StarIcon />}
        />
        <IconButton
          data-testid={`city-card-wishlist-btn${visited ? '-active' : ''}`}
          mx={1}
          variant="outline"
          colorScheme={wishlist ? 'blue' : 'red'}
          title={wishlistBtnLabel}
          aria-label={wishlistBtnLabel}
          onClick={onToggleWishlist}
          icon={<PlusSquareIcon />}
        />
      </Box>
    </Box>
  )
}
