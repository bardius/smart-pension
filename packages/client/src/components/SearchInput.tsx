import { Box, IconButton, Input, InputGroup, InputRightElement } from '@chakra-ui/react'
import { Search2Icon } from '@chakra-ui/icons'
import { useCallback, useState } from 'react'
import type { VFC } from 'react'

export type SearchInputProps = {
  onChange: (value: string) => void
  onSubmit: (value: string) => void
}

export const SearchInput: VFC<SearchInputProps> = ({ onChange, onSubmit }) => {
  const [searchQuery, setSearchQuery] = useState('')

  const onSearchChange = useCallback(
    event => {
      const newSearchQuery = event.target.value || ''
      setSearchQuery(newSearchQuery)
      onChange(newSearchQuery)
    },
    [onChange, setSearchQuery]
  )

  const onSearchSubmit = useCallback(() => {
    onSubmit(searchQuery)
  }, [searchQuery, onSubmit])

  return (
    <Box as="form" role="search">
      <InputGroup>
        <Input
          data-testid="search-input"
          onChange={onSearchChange}
          value={searchQuery}
          type="search"
          aria-label={'Search by city name, search results will appear below'}
        />
        <InputRightElement
          data-testid="search-input-btn"
          onClick={onSearchSubmit}
          children={
            <IconButton aria-label={`Search cities by name containing ${searchQuery}`} icon={<Search2Icon />} />
          }
        />
      </InputGroup>
    </Box>
  )
}
