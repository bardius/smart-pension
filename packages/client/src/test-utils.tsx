import * as React from 'react'
import type { RenderOptions } from '@testing-library/react'
import { render } from '@testing-library/react'
import { ChakraProvider, theme } from '@chakra-ui/react'
import { MockedProvider } from '@apollo/client/testing'
import { citiesQueriesMocks } from './__mocks__/citiesQueries'

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <MockedProvider mocks={citiesQueriesMocks} addTypename={false}>
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  </MockedProvider>
)

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options })

export { customRender as render }
