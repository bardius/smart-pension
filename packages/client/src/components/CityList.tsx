import React from 'react'
import type { VFC } from 'react'
import { SimpleGrid } from '@chakra-ui/react'
import type { City } from '../domain/cities/types'
import { CityCard } from './CityCard'

export type CityListProps = {
  cities?: City[]
}

// We could add aria-live="polite" aria-relevant="additions removals" in SimpleGrid but maybe it is to much for this list
export const CityList: VFC<CityListProps> = props => {
  const { cities } = props

  return (
    <>
      {cities?.length === 0 && <p data-testid={'city-list-empty'}>No results</p>}

      {cities && cities.length > 0 && (
        <SimpleGrid as="ul" columns={1} spacing={10} width={'100%'} mb={10} data-testid={'city-list'}>
          {cities.map(city => {
            return (
              <CityCard
                key={city.id}
                id={city.id}
                title={city.name}
                subTitle={city.country}
                visited={city.visited}
                wishlist={city.wishlist}
              />
            )
          })}
        </SimpleGrid>
      )}
    </>
  )
}
