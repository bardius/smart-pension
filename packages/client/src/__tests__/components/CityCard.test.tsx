import React from 'react'
import { fireEvent, screen } from '@testing-library/react'
import { CityCard } from '../../components/CityCard'
import { render } from '../../test-utils'

const mockUpdateCityFn = jest.fn()

jest.mock('../../contexts/cities/hooks', () => ({
  useUpdateCity: () => mockUpdateCityFn,
}))

describe('Component - CityCard', () => {
  it('Should render city with visited & wishlist selection', () => {
    const renderLayout = render(
      <CityCard id={1} title={'London'} subTitle={'United Kingdom'} visited={true} wishlist={true} />
    )

    expect(screen.getByTestId('city-card')).toBeInTheDocument()
    expect(screen.getByTestId('city-card-wishlist-btn-active')).toBeInTheDocument()
    expect(screen.getByTestId('city-card-visited-btn-active')).toBeInTheDocument()

    const { asFragment } = renderLayout
    expect(asFragment()).toMatchSnapshot()
  })

  it('Should render city without visited & wishlist selection', () => {
    const renderLayout = render(
      <CityCard id={1} title={'London'} subTitle={'United Kingdom'} visited={false} wishlist={false} />
    )
    expect(screen.getByTestId('city-card')).toBeInTheDocument()
    expect(screen.getByTestId('city-card-wishlist-btn')).toBeInTheDocument()
    expect(screen.getByTestId('city-card-visited-btn')).toBeInTheDocument()

    const { asFragment } = renderLayout
    expect(asFragment()).toMatchSnapshot()
  })

  it('Should make request to toggle visited state', () => {
    render(<CityCard id={1} title={'London'} subTitle={'United Kingdom'} visited={false} wishlist={false} />)

    const visitedBtn = screen.getByTestId('city-card-visited-btn')

    fireEvent.click(visitedBtn)

    expect(mockUpdateCityFn).toHaveBeenCalledTimes(1)
  })

  it('Should make request to toggle wishlist state', () => {
    render(<CityCard id={1} title={'London'} subTitle={'United Kingdom'} visited={false} wishlist={false} />)

    const wishlistBtn = screen.getByTestId('city-card-wishlist-btn')

    fireEvent.click(wishlistBtn)

    expect(mockUpdateCityFn).toHaveBeenCalledTimes(1)
  })
})
