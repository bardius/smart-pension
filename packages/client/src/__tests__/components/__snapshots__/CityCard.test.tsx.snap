// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - CityCard Should render city with visited & wishlist selection 1`] = `
<DocumentFragment>
  <li
    class="css-16ax18g"
    data-testid="city-card"
  >
    <div
      class="css-qiuk2s"
    >
      <p
        class="chakra-text css-wjomy9"
      >
        London
      </p>
      <span
        class="chakra-text css-3b0vjl"
      >
        (United Kingdom)
      </span>
    </div>
    <div
      class="css-zdpt2t"
    >
      <button
        aria-label="Marked as visited, Click to toggle status"
        class="chakra-button css-8x7hex"
        data-testid="city-card-visited-btn-active"
        title="Marked as visited, Click to toggle status"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M23.555,8.729a1.505,1.505,0,0,0-1.406-.98H16.062a.5.5,0,0,1-.472-.334L13.405,1.222a1.5,1.5,0,0,0-2.81,0l-.005.016L8.41,7.415a.5.5,0,0,1-.471.334H1.85A1.5,1.5,0,0,0,.887,10.4l5.184,4.3a.5.5,0,0,1,.155.543L4.048,21.774a1.5,1.5,0,0,0,2.31,1.684l5.346-3.92a.5.5,0,0,1,.591,0l5.344,3.919a1.5,1.5,0,0,0,2.312-1.683l-2.178-6.535a.5.5,0,0,1,.155-.543l5.194-4.306A1.5,1.5,0,0,0,23.555,8.729Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Added in wishlist, Click to toggle status"
        class="chakra-button css-8x7hex"
        data-testid="city-card-wishlist-btn-active"
        title="Added in wishlist, Click to toggle status"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-width="2"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="3"
            />
            <path
              d="M12 8v8"
            />
            <path
              d="M8 12h8"
            />
          </g>
        </svg>
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`Component - CityCard Should render city without visited & wishlist selection 1`] = `
<DocumentFragment>
  <li
    class="css-16ax18g"
    data-testid="city-card"
  >
    <div
      class="css-qiuk2s"
    >
      <p
        class="chakra-text css-wjomy9"
      >
        London
      </p>
      <span
        class="chakra-text css-3b0vjl"
      >
        (United Kingdom)
      </span>
    </div>
    <div
      class="css-zdpt2t"
    >
      <button
        aria-label="Marked as not visited, Click to toggle status"
        class="chakra-button css-1o56cth"
        data-testid="city-card-visited-btn"
        title="Marked as not visited, Click to toggle status"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M23.555,8.729a1.505,1.505,0,0,0-1.406-.98H16.062a.5.5,0,0,1-.472-.334L13.405,1.222a1.5,1.5,0,0,0-2.81,0l-.005.016L8.41,7.415a.5.5,0,0,1-.471.334H1.85A1.5,1.5,0,0,0,.887,10.4l5.184,4.3a.5.5,0,0,1,.155.543L4.048,21.774a1.5,1.5,0,0,0,2.31,1.684l5.346-3.92a.5.5,0,0,1,.591,0l5.344,3.919a1.5,1.5,0,0,0,2.312-1.683l-2.178-6.535a.5.5,0,0,1,.155-.543l5.194-4.306A1.5,1.5,0,0,0,23.555,8.729Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Not in wishlist, Click to toggle status"
        class="chakra-button css-1o56cth"
        data-testid="city-card-wishlist-btn"
        title="Not in wishlist, Click to toggle status"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-width="2"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="3"
            />
            <path
              d="M12 8v8"
            />
            <path
              d="M8 12h8"
            />
          </g>
        </svg>
      </button>
    </div>
  </li>
</DocumentFragment>
`;
